smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGenerator.kt

```diff
@@ -282,8 +280,33 @@ abstract class HttpProtocolClientGenerator(
      * Render any additional methods to support client operation
      */
     protected open fun renderAdditionalMethods(writer: KotlinWriter) { }
-}
 
-// TODO https://github.com/awslabs/aws-sdk-kotlin/issues/557
-private fun OperationShape.checksumRequired(): Boolean =
-    hasTrait<HttpChecksumRequiredTrait>() || getTrait<HttpChecksumTrait>()?.isRequestChecksumRequired == true
+    /**
+     * Render optionally installing Md5ChecksumMiddleware.
+     * The Md5 middleware will only be installed if the operation requires a checksum and the user has not opted-in to flexible checksums.
+     */
+    private fun OperationShape.renderIsMd5ChecksumRequired(writer: KotlinWriter) {
```

Comment:
**nit**: It feels odd to have a `render` method off the operation shape. Also this could probably just be turned into a [default middleware](https://github.com/awslabs/smithy-kotlin/blob/f29e7502abc4dd817fcf2338e742e67ad1ecb498/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt#L64) such that `HttpProtocolClientGenerator` need not know anything about it.